File: crates/rustakka-profiler/src/report.rs
//! Shared report schema.
//!
//! The same structure is produced by the Python profiler so the two
//! runtimes can be merged into a single comparison table.

use std::time::Duration;

use serde::{Deserialize, Serialize};

/// Which workload was measured.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Scenario {
    /// Fire-and-forget `tell` into a null actor.
    Tell,
    /// Sequential `ask` round-trips (measures latency).
    Ask,
    /// Spawn many actors and hit each one once (actor-creation cost).
    Fanout,
    /// CPU-bound handler (xxHash-lite compute loop).
    Cpu,
}

impl Scenario {
    pub fn name(self) -> &'static str {
        match self {
            Scenario::Tell => "tell",
            Scenario::Ask => "ask",
            Scenario::Fanout => "fanout",
            Scenario::Cpu => "cpu",
        }
    }

    pub fn parse(s: &str) -> Option<Self> {
        match s {
            "tell" => Some(Scenario::Tell),
            "ask" => Some(Scenario::Ask),
            "fanout" => Some(Scenario::Fanout),
            "cpu" => Some(Scenario::Cpu),
            _ => None,
        }
    }

    pub fn all() -> &'static [Scenario] {
        &[Scenario::Tell, Scenario::Ask, Scenario::Fanout, Scenario::Cpu]
    }
}

/// One scenario's measurement. Times are in nanoseconds for precision;
/// helpers render them as human-friendly units.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Measurement {
    pub runtime: String,          // "rust" or "python"
    pub scenario: Scenario,
    pub config: String,           // free-form (dispatcher, pool size, ...)
    pub messages: u64,            // total messages (or actors for fanout)
    pub elapsed_ns: u64,
    pub throughput_msgs_per_sec: f64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub p50_ns: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub p95_ns: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub p99_ns: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub rss_delta_bytes: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub peak_rss_bytes: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cpu_delta_ns: Option<u64>,
}

impl Measurement {
    pub fn from_throughput(
        runtime: &str,
        scenario: Scenario,
        config: &str,
        messages: u64,
        elapsed: Duration,
    ) -> Self {
        let elapsed_ns = elapsed.as_nanos() as u64;
        let throughput = if elapsed_ns == 0 {
            0.0
        } else {
            (messages as f64) * 1.0e9 / (elapsed_ns as f64)
        };
        Self {
            runtime: runtime.to_string(),
            scenario,
            config: config.to_string(),
            messages,
            elapsed_ns,
            throughput_msgs_per_sec: throughput,
            p50_ns: None,
            p95_ns: None,
            p99_ns: None,
            rss_delta_bytes: None,
            peak_rss_bytes: None,
            cpu_delta_ns: None,
        }
    }

    pub fn with_latencies(mut self, sorted: &[Duration]) -> Self {
        use crate::metrics::percentile;
        self.p50_ns = percentile(sorted, 50.0).map(|d| d.as_nanos() as u64);
        self.p95_ns = percentile(sorted, 95.0).map(|d| d.as_nanos() as u64);
        self.p99_ns = percentile(sorted, 99.0).map(|d| d.as_nanos() as u64);
        self
    }

    pub fn with_memory(mut self, delta: Option<i64>, peak: Option<u64>) -> Self {
        self.rss_delta_bytes = delta;
        self.peak_rss_bytes = peak;
        self
    }

    pub fn with_cpu(mut self, cpu: Option<Duration>) -> Self {
        self.cpu_delta_ns = cpu.map(|d| d.as_nanos() as u64);
        self
    }
}

/// Top-level report — a list of measurements plus some environment metadata.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProfilerReport {
    pub runtime: String,
    pub version: String,
    pub host: String,
    pub measurements: Vec<Measurement>,
}

impl ProfilerReport {
    pub fn new(runtime: &str) -> Self {
        Self {
            runtime: runtime.to_string(),
            version: env!("CARGO_PKG_VERSION").to_string(),
            host: host_tag(),
            measurements: Vec::new(),
        }
    }

    pub fn push(&mut self, m: Measurement) {
        self.measurements.push(m);
    }

    /// Render as a human-friendly markdown table.
    pub fn to_markdown(&self) -> String {
        let mut out = String::new();
        out.push_str(&format!(
            "# rustakka profiler — {} ({})\n\nhost: `{}`\n\n",
            self.runtime, self.version, self.host
        ));
        out.push_str(
            "| scenario | config | msgs | elapsed | throughput | p50 | p95 | p99 | ΔRSS | CPU |\n",
        );
        out.push_str("|---|---|---|---|---|---|---|---|---|---|\n");
        for m in &self.measurements {
            out.push_str(&format!(
                "| {} | {} | {} | {} | {} | {} | {} | {} | {} | {} |\n",
                m.scenario.name(),
                m.config,
                m.messages,
                fmt_ns(m.elapsed_ns),
                fmt_rate(m.throughput_msgs_per_sec),
                fmt_opt_ns(m.p50_ns),
                fmt_opt_ns(m.p95_ns),
                fmt_opt_ns(m.p99_ns),
                fmt_opt_delta(m.rss_delta_bytes),
                fmt_opt_ns(m.cpu_delta_ns),
            ));
        }
        out
    }
}

fn host_tag() -> String {
    let os = std::env::consts::OS;
    let arch = std::env::consts::ARCH;
    let cpus = std::thread::available_parallelism()
        .map(|n| n.get())
        .unwrap_or(0);
    format!("{os}/{arch} cpus={cpus}")
}

fn fmt_ns(ns: u64) -> String {
    if ns >= 1_000_000_000 {
        format!("{:.2}s", ns as f64 / 1e9)
    } else if ns >= 1_000_000 {
        format!("{:.2}ms", ns as f64 / 1e6)
    } else if ns >= 1_000 {
        format!("{:.2}µs", ns as f64 / 1e3)
    } else {
        format!("{ns}ns")
    }
}

fn fmt_opt_ns(v: Option<u64>) -> String {
    match v {
        Some(n) => fmt_ns(n),
        None => "n/a".to_string(),
    }
}

fn fmt_rate(v: f64) -> String {
    if v >= 1e6 {
        format!("{:.2}M/s", v / 1e6)
    } else if v >= 1e3 {
        format!("{:.2}k/s", v / 1e3)
    } else {
        format!("{v:.2}/s")
    }
}

fn fmt_opt_delta(v: Option<i64>) -> String {
    match v {
        Some(n) => {
            let abs = n.unsigned_abs();
            let pretty = if abs >= 1 << 30 {
                format!("{:.2}GiB", abs as f64 / (1u64 << 30) as f64)
            } else if abs >= 1 << 20 {
                format!("{:.2}MiB", abs as f64 / (1u64 << 20) as f64)
            } else if abs >= 1 << 10 {
                format!("{:.2}KiB", abs as f64 / (1u64 << 10) as f64)
            } else {
                format!("{abs}B")
            };
            if n < 0 { format!("-{pretty}") } else { format!("+{pretty}") }
        }
        None => "n/a".to_string(),
    }
}

File: python/atomr/profiler/_report.py
"""Dataclasses + table formatting for the Python profiler.

Kept schema-compatible with ``atomr-profiler`` so reports can be
merged across runtimes.
"""

from __future__ import annotations

import json
from dataclasses import asdict, dataclass, field
from typing import List, Optional


@dataclass
class Measurement:
    runtime: str
    scenario: str
    config: str
    messages: int
    elapsed_ns: int
    throughput_msgs_per_sec: float
    p50_ns: Optional[int] = None
    p95_ns: Optional[int] = None
    p99_ns: Optional[int] = None
    rss_delta_bytes: Optional[int] = None
    peak_rss_bytes: Optional[int] = None
    cpu_delta_ns: Optional[int] = None

    def to_dict(self) -> dict:
        return {k: v for k, v in asdict(self).items() if v is not None}


@dataclass
class ProfilerReport:
    runtime: str = "python"
    version: str = ""
    host: str = ""
    measurements: List[Measurement] = field(default_factory=list)

    def to_json(self) -> str:
        return json.dumps(
            {
                "runtime": self.runtime,
                "version": self.version,
                "host": self.host,
                "measurements": [m.to_dict() for m in self.measurements],
            },
            indent=2,
        )

    def to_markdown(self) -> str:
        rows = [
            f"# atomr profiler — {self.runtime} ({self.version})",
            "",
            f"host: `{self.host}`",
            "",
            "| scenario | config | msgs | elapsed | throughput | p50 | p95 | p99 | ΔRSS | CPU |",
            "|---|---|---|---|---|---|---|---|---|---|",
        ]
        for m in self.measurements:
            rows.append(
                "| {s} | {c} | {n} | {e} | {t} | {p50} | {p95} | {p99} | {rss} | {cpu} |".format(
                    s=m.scenario,
                    c=m.config,
                    n=m.messages,
                    e=fmt_ns(m.elapsed_ns),
                    t=fmt_rate(m.throughput_msgs_per_sec),
                    p50=fmt_opt_ns(m.p50_ns),
                    p95=fmt_opt_ns(m.p95_ns),
                    p99=fmt_opt_ns(m.p99_ns),
                    rss=fmt_opt_delta(m.rss_delta_bytes),
                    cpu=fmt_opt_ns(m.cpu_delta_ns),
                )
            )
        return "\n".join(rows) + "\n"


def fmt_ns(ns: int) -> str:
    if ns >= 1_000_000_000:
        return f"{ns / 1e9:.2f}s"
    if ns >= 1_000_000:
        return f"{ns / 1e6:.2f}ms"
    if ns >= 1_000:
        return f"{ns / 1e3:.2f}µs"
    return f"{ns}ns"


def fmt_opt_ns(v: Optional[int]) -> str:
    return fmt_ns(v) if v is not None else "n/a"


def fmt_rate(v: float) -> str:
    if v >= 1e6:
        return f"{v / 1e6:.2f}M/s"
    if v >= 1e3:
        return f"{v / 1e3:.2f}k/s"
    return f"{v:.2f}/s"


def fmt_opt_delta(v: Optional[int]) -> str:
    if v is None:
        return "n/a"
    sign = "-" if v < 0 else "+"
    n = abs(v)
    if n >= 1 << 30:
        return f"{sign}{n / (1 << 30):.2f}GiB"
    if n >= 1 << 20:
        return f"{sign}{n / (1 << 20):.2f}MiB"
    if n >= 1 << 10:
        return f"{sign}{n / (1 << 10):.2f}KiB"
    return f"{sign}{n}B"


def percentile(sorted_vals: List[int], p: float) -> Optional[int]:
    if not sorted_vals:
        return None
    idx = min(len(sorted_vals) - 1, max(0, round((p / 100.0) * (len(sorted_vals) - 1))))
    return sorted_vals[idx]
